feat: api rabbitmq connector

raduachim opened the pull request:
Adds infrastructure for using rabbitmq from api to communicate with engine
Implements create booking route

BREAKING CHANGE:
The engine no longer generates a booking id, one should generate a booking id and include it in the booking message in the format pmb-[8 characters from alphabet]

Co-authored-by: Adam Näslund iteadam@users.noreply.github.com
Co-authored-by: Emil Ahlbäck mull@users.noreply.github.com

Commit:
chore: WIP on adding rabbitmq functionality in api

added jest part, copied engineAdapter and rabbitmqConnector, needs tests

Co-authored-by: Adam Näslund <iteadam@users.noreply.github.com>
Co-authored-by: Emil Ahlbäck <mull@users.noreply.github.com>

Commit:
chore: added rabbitmq connector and engine adapter

Commit:
chore: WIP add tests

Commit:
chore: "refactoring" code

we made some types, removed generic stream and added a specific case for create

Co-authored-by: Emil Ahlbäck <mull@users.noreply.github.com>
Co-authored-by: Adam Näslund <iteadam@users.noreply.github.com>

iteadam committed:
Work on the engineAdapter, also integration tests!

Commit:
test: 💍 fix engine tests after removing id from Booking.make

also removed assigned_to from Booking.make since it isn't used in the
function

mull committed:
wip: endpoint work

Commit:
chore: event emitter for new booking

forking streams every time when creating a booking is dumb

Co-authored-by: Emil Ahlbäck <mull@users.noreply.github.com>
Co-authored-by: Adam Näslund <iteadam@users.noreply.github.com>

mull commented on packages/api/src/routeHandlers.ts:
```diff
@@ -9,8 +18,8 @@ export const get_transports: Handler = (c, req, res) => {
       },
-      earliest_start: '',
-      latest_end: '',
+      earliest_start: '2018-01-01 14:10:10',
+      latest_end: 'tji',
```
```suggestion
      earliest_start: '2021-07-01T17:30:00Z',
      latest_end: ''2021-07-01T21:30:00Z',
```

mull committed:
Update packages/api/src/routeHandlers.ts

Commit:
chore: cleanup

Co-authored-by: Adam Näslund <iteadam@users.noreply.github.com>

raduachim commented on packages/engine_umbrella/apps/message_generator/mix.exs:
```diff
@@ -31,6 +31,7 @@ defmodule MessageGenerator.MixProject do
       {:httpoison, "~> 1.8"},
       {:amqp, "~> 1.4"},
+      {:elixir_uuid, "~> 1.2"},
       # {:hackney, git: "https://github.com/benoitc/hackney.git", branch: "master", override: true}
```
```suggestion
```
Remove if commented

raduachim committed:
Update packages/engine_umbrella/apps/message_generator/mix.exs

Commit:
fix: api subscribe to all booking updates topics

Commit:
chore: 🤖 remove api from docker-compose file

since it's going through constant development changes it is easier to
run it locally and when it's stable it should be added back with the
same way of running it with volumes

Commit:
chore: 🤖 use one connection for rabbitmq

instead of creating a new connection every time we would use amqp,
better use one connection and let it create new channels instead on that
connection; also fixing jest process to exit when it finished running
tests
